feat: ✨ automatically replace invalid enum expressions with corresponding valid expression & import

ringohoffman opened the pull request:
Inspired by: https://github.com/sizmailov/pybind11-stubgen/issues/192

I think this feature supersedes the need for `--enum-class-locations`, though I did not remove it here. At least that is my goal. I want this project to be as easy for people to pick up as possible, and this was something I thought should be possible for `pybind11_stubgen` to do on behalf of the user.

We tweak `RewritePybind11EnumValueRepr` to delay `report_error()` for enum expressions. We remember all of the invalid enum expressions we have seen as we crawl through the module. When we find the definition of a value corresponding to an enum repr, we use it to fix invalid expressions we have seen or will see. If there are still invalid enum expressions left over, we `report_error()` during `finalize()`.

sizmailov commented:
This should eliminate the need for enum location specification for most use cases.

This PR introduces a bit of guessing to stub generation if the module has two enums with the same class and value names. Depending on the order of traversing the module, the resulting stubs could be different.

I suggest the following changes:
- Defer all the substitutions to the very end to catch such ambiguous replacements
- Not replace ambiguous enum values, e.g. `foo(x: Any = <Color.RED: 3>)`
- Replace unambiguous enums (e.g. `foo(x: my_module.Car.Color = <Color.RED: 3>)`
- Do not import values into the module.

sizmailov commented on tests/stubs/python-3.12/pybind11-master/numpy-array-use-type-var/demo/_bindings/aliases/__init__.pyi:
```diff
@@ -5,9 +5,9 @@ import typing
-import demo._bindings.enum
 from demo._bindings.aliases.foreign_method_arg import Bar2 as foreign_type_alias
 from demo._bindings.aliases.foreign_return import get_foo as foreign_class_alias
+from demo._bindings.enum import ConsoleForegroundColor
```
Let's not do value imports. They are problematic in a couple of ways.

For example, this would make the following code pass a static analyzer but fail at runtime.

```python 
from demo._bindings.aliases import ConsoleForegroundColor
```

ringohoffman commented:
@sizmailov are you planning to drop support for 3.7 at some point? [It reached end of life status 5 months ago](https://devguide.python.org/versions/). I would like it if we did, because I want to use the assignment expression operator (`:=`). I think it can be great for readability if used correctly.

[3.8's EOL is coming in only 10 months](https://endoflife.date/python), also.  [numpy](https://github.com/numpy/numpy/blob/94bc56402a99c85c72330ca54c5f1f4bc2aac9f7/pyproject.toml#L19) has already dropped support for it in preparation.

What are your thoughts about this?

sizmailov commented:
I think we should aim for the same versions as pybind11, which is currently 3.6+.
I don't remember exactly why I didn't support 3.6. Probably, it was the [troubles of setting up the test for CI ](https://github.com/sizmailov/pybind11-stubgen/actions/runs/6034143776/job/16371984393) :disappointed: 

So Python 3.7 support would stay for quite a while, for better or worse.

ringohoffman commented:
@sizmailov Okay, I gave it my best effort.

1. I moved the fixing to `finalize()`.
2. I check to see if there are multiple possible definitions for an invalid enum repr, and raise an error like:
```
pybind11_stubgen - [  ERROR] In finalize : Enum member '<ConsoleForegroundColor.Magenta: 35>' could not be resolved; multiple identical definitions found in: 'demo._bindings.duplicate_enum', 'demo._bindings.enum'
```
3. [I added a new test that creates an ambiguous enum using `ConsoleForegroundColorDuplicate`](https://github.com/sizmailov/pybind11-stubgen/pull/196/files#diff-afe80135a264db883f315b534da7b5b04af9aab383be19251e2895a5536ab42cR7-R14). I had to update `--ignore-invalid-expressions` to incldue `<ConsoleForegroundColor\\.Magenta: 35>`.
4. I needed to add a `LoggerData` override of `finalize()` to create a new layer so that I could call `report_error()` from `finalize()`.

sizmailov commented on pybind11_stubgen/parser/errors.py:
```diff
@@ -33,3 +33,23 @@ def __init__(self, name: QualifiedName):
+        origins = sorted(import_.origin for _, import_ in self.values_and_imports)
+        return (
+            f"Enum member '{self.expression}' could not be resolved; multiple "
+            "identical definitions found in: "
```
```suggestion
            "matching definitions found in: "
```

sizmailov commented on pybind11_stubgen/parser/mixins/fix.py:
```diff
@@ -999,13 +1001,56 @@ def parse_value_str(self, value: str) -> Value | InvalidExpression:
+    2. The `InvalidExpression` repr corresponds to multiple enum field definitions. An
+    `AmbiguousEnumError` is reported.
+    3. The `InvalidExpression` repr corresponds to no enum field definitions. An
+    `InvalidEnumError` is reported.
```
```suggestion
    `InvalidExpressionError` is reported.
```

sizmailov commented on tests/demo.errors.stderr.txt:
```diff
@@ -1,19 +1,15 @@
-pybind11_stubgen - [WARNING] Enum-like str representations were found with no matching mapping to the enum class location.
-Use `--enum-class-locations` to specify full path to the following enum(s):
- - ConsoleForegroundColor
+pybind11_stubgen - [  ERROR] In finalize : Enum member '<ConsoleForegroundColor.Magenta: 35>' could not be resolved; multiple identical definitions found in: 'demo._bindings.duplicate_enum', 'demo._bindings.enum'
```
```suggestion
pybind11_stubgen - [  ERROR] In finalize : Enum member '<ConsoleForegroundColor.Magenta: 35>' could not be resolved; multiple matching definitions found in: 'demo._bindings.duplicate_enum', 'demo._bindings.enum'
```

sizmailov reviewed (commented):
Thank you!

The change looks good to me.

Just a couple of doc suggestions.

sizmailov commented:
If you don't mind, I'll publish it as 2.4.1

ringohoffman commented:
I don't mind! Thanks you for reviewing my work!
